// bilibili-live-ws/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KeepLiveTCP = exports.KeepLiveWS = exports.LiveTCP = exports.LiveWS = exports.relayEvent = exports.getRoomid = exports.getConf = void 0;
const node_1 = require("./inflate/node");
const ws_1 = require("./ws");
const tcp_1 = require("./tcp");
const common_1 = require("./common");
var extra_1 = require("./extra");
Object.defineProperty(exports, "getConf", { enumerable: true, get: function () { return extra_1.getConf; } });
Object.defineProperty(exports, "getRoomid", { enumerable: true, get: function () { return extra_1.getRoomid; } });
var common_2 = require("./common");
Object.defineProperty(exports, "relayEvent", { enumerable: true, get: function () { return common_2.relayEvent; } });
class LiveWS extends ws_1.LiveWSBase {
    constructor(roomid, opts) {
        super(node_1.inflates, roomid, opts);
    }
}
exports.LiveWS = LiveWS;
class LiveTCP extends tcp_1.LiveTCPBase {
    constructor(roomid, opts) {
        super(node_1.inflates, roomid, opts);
    }
}
exports.LiveTCP = LiveTCP;
class KeepLiveWS extends common_1.KeepLive {
    constructor(roomid, opts) {
        super(ws_1.LiveWSBase, node_1.inflates, roomid, opts);
    }
}
exports.KeepLiveWS = KeepLiveWS;
class KeepLiveTCP extends common_1.KeepLive {
    constructor(roomid, opts) {
        super(tcp_1.LiveTCPBase, node_1.inflates, roomid, opts);
    }
}
exports.KeepLiveTCP = KeepLiveTCP;
