// bilibili-live-ws/src/ws.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiveWSBase = exports.isNode = void 0;
const events_1 = require("events");
const isomorphic_ws_1 = __importDefault(require("isomorphic-ws"));
const common_1 = require("./common");
exports.isNode = !!isomorphic_ws_1.default.Server;
class WebSocket extends events_1.EventEmitter {
    constructor(address, inflates, ...args) {
        super();
        const ws = new isomorphic_ws_1.default(address, ...(exports.isNode ? args : []));
        this.ws = ws;
        ws.onopen = () => this.emit('open');
        ws.onmessage = exports.isNode ? ({ data }) => this.emit('message', data) : async ({ data }) => this.emit('message', inflates.Buffer.from(await new Response(data).arrayBuffer()));
        ws.onerror = () => this.emit('error');
        ws.onclose = () => this.emit('close');
    }
    get readyState() {
        return this.ws.readyState;
    }
    send(data) {
        this.ws.send(data);
    }
    close(code, data) {
        this.ws.close(code, data);
    }
}
class LiveWSBase extends common_1.Live {
    constructor(inflates, roomid, { address = 'wss://broadcastlv.chat.bilibili.com/sub', agent, ...options } = {}) {
        const ws = new WebSocket(address, inflates, { agent });
        const send = (data) => {
            if (ws.readyState === 1) {
                ws.send(data);
            }
        };
        const close = () => this.ws.close();
        super(inflates, roomid, { send, close, ...options });
        ws.on('open', (...params) => this.emit('open', ...params));
        ws.on('message', data => this.emit('message', data));
        ws.on('close', (code, reason) => this.emit('close', code, reason));
        ws.on('error', error => this.emit('_error', error));
        this.ws = ws;
    }
}
exports.LiveWSBase = LiveWSBase;
